### PR TITLE
[CI/CD] Update release workflow and introduce workflow for nightly releases

### PR DESCRIPTION
**Pull request by alexander-smolyakov**
resolves - N/A

### Description

This pull request contains the following changes:
- Updates for release workflow to use reusable workflows;
- Introduces nightly releases workflow;

**Changelog**:
_release.yml_:
- Added `workflow_call` trigger;
- Updated workflow inputs;
- Steps replaced with reusable workflows;

_nightly-release.yml_:
- Added initial implementation;

_env-setup.sh_:
- Added script to setup environment for integration tests;

_.bumpversion.cfg_:
- Updated `parse` and `serialize` properties of config to respect the development segment for nightly releases;

### Checklist

- [x] I have read [the contributing guide](https://github.com/dbt-labs/dbt-core/blob/main/CONTRIBUTING.md) and understand what's expected of me
- [x] I have signed the [CLA](https://docs.getdbt.com/docs/contributor-license-agreements)
- [ ] I have run this code in development and it appears to resolve the stated issue
- [x] This PR includes tests, or tests are not required/relevant for this PR
- [x] I have [opened an issue to add/update docs](https://github.com/dbt-labs/docs.getdbt.com/issues/new/choose), or docs changes are not required/relevant for this PR
- [ ] I have run `changie new` to [create a changelog entry](https://github.com/dbt-labs/dbt-core/blob/main/CONTRIBUTING.md#adding-a-changelog-entry)

### CODE REVIEW

File: .github/workflows/nightly-release.yml [hunk 1]
```diff
@@ -0,0 +1,106 @@
+# **what?**
+# Nightly releases to GitHub and PyPI. This workflow produce the following outcome:
```
Comment (emmyoop) [inline .github/workflows/nightly-release.yml:2]:
```suggestion
# Nightly releases to GitHub and PyPI. This workflow produces the following outcome:
```

File: .github/workflows/nightly-release.yml [hunk 2]
```diff
@@ -0,0 +1,106 @@
...
+# - generate and validate data for night release (commit SHA, version number, release branch);
+# - pass data to release workflow;
+# - night build will be pushed to test PyPI;
+# - night release will be pushed to GitHub as a draft release;
```
Comment (emmyoop) [inline .github/workflows/nightly-release.yml:6]: Let's keep these in the order they happen
```suggestion
# - night release will be pushed to GitHub as a draft release;
# - night build will be pushed to test PyPI;
```

File: .github/workflows/nightly-release.yml [hunk 3]
```diff
@@ -0,0 +1,106 @@
...
+#
+# **why?**
+# Ensure an automated and tested release process for nightly builds
+#
+# **when?**
+# This workflow runs on schedule or can be run manually on demand.
+
+name: Nightly Release to GitHub and PyPi
```
Comment (emmyoop) [inline .github/workflows/nightly-release.yml:14]:
```suggestion
name: Nightly Test Release to GitHub and PyPi
```

**Review by emmyoop — CHANGES_REQUESTED** [review]

File: .github/workflows/nightly-release.yml (hunk 3 above)
Comment (alexander-smolyakov) [reply .github/workflows/nightly-release.yml]: Fixed in https://github.com/dbt-labs/dbt-core/pull/6602/commits/2fbde2cc196d0f942da42cb03b54b25be4e95e53

File: .github/workflows/nightly-release.yml (hunk 2 above)
Comment (alexander-smolyakov) [reply .github/workflows/nightly-release.yml]: Fixed in https://github.com/dbt-labs/dbt-core/pull/6602/commits/2fbde2cc196d0f942da42cb03b54b25be4e95e53

File: .github/workflows/nightly-release.yml (hunk 1 above)
Comment (alexander-smolyakov) [reply .github/workflows/nightly-release.yml]: Fixed in https://github.com/dbt-labs/dbt-core/pull/6602/commits/2fbde2cc196d0f942da42cb03b54b25be4e95e53

File: .github/workflows/release.yml [hunk 4]
```diff
@@ -1,24 +1,110 @@
...
+        required: true
+      package_test_command:
+        description: "Package test command"
+        type: string
+        default: "dbt --version"
+        required: true
+      test_run:
+        description: "Test run (Publish release as draft)"
+        type: boolean
+        default: false
```
[the remark this answers is not in the document]
Comment (alexander-smolyakov) [reply .github/workflows/release.yml]: Fixed in https://github.com/dbt-labs/dbt-core/pull/6602/commits/2fbde2cc196d0f942da42cb03b54b25be4e95e53

File: .github/workflows/release.yml [hunk 5]
```diff
@@ -28,175 +114,117 @@ defaults:
...
-        with:
-          name: dist
-          path: 'dist'
-
-      - name: Publish distribution to PyPI
-        uses: pypa/gh-action-pypi-publish@v1.4.2
-        with:
-          password: ${{ secrets.PYPI_API_TOKEN }}
+    secrets:
+      SLACK_WEBHOOK_URL: ${{ secrets.SLACK_TEST_CHANNEL }}
```
Comment (emmyoop) [inline .github/workflows/release.yml:230]:
```suggestion
      SLACK_WEBHOOK_URL: ${{ secrets.SLACK_DEV_CORE_ALERTS }}
```

**Review by emmyoop — APPROVED** [review]
This looks good.  We won't be setting up the temp slack channel in this repo so we need to update the token to be the final token.

**Review by emmyoop — APPROVED** [review]

**Pull request merged**